{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9540eedc",
   "metadata": {},
   "source": [
    "# Notebook with working models in NumPyro"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2778dd2b",
   "metadata": {},
   "source": [
    "--------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cf98a949",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load necessary packages\n",
    "import numpyro\n",
    "import numpyro.distributions as dist\n",
    "\n",
    "import jax\n",
    "import jax.numpy as jnp\n",
    "\n",
    "from helper import custom_put_along_axis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "ea1921cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Minimal example\n",
    "reference = jnp.array([0])\n",
    "reads = jnp.array([[0], [1], [1], [1], [0], [1], [0], [1]])\n",
    "alphabet ='01'\n",
    "\n",
    "cluster_num = 5\n",
    "\n",
    "input_data = reference, reads, len(alphabet), cluster_num"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "744d2418",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0886777d",
   "metadata": {},
   "source": [
    "## Model - Finite DPM\n",
    "\n",
    "#### Finite Dirichlet mixture model with fixed number of clusters. \n",
    "\n",
    "ref = reference sequence  \n",
    "$N$ = number of reads  \n",
    "$K$ = number of clusters/components  \n",
    "$L$ = genome length (number of positions)  \n",
    "alphabet = {A, C, G, T, -}\n",
    "\n",
    "\n",
    "no-mutation rate: $\\gamma \\sim Beta(a,b)$   \n",
    "no-error rate: $\\theta \\sim Beta(c,d)$   \n",
    "Cluster weights ($K$-dim): $\\pi | \\alpha \\sim Dir(\\alpha)$  \n",
    "Cluster assignments ($N$-dim): $z|\\pi \\sim Categorical(\\pi)$  \n",
    "Cluster centers/haplotypes ($K$x$L$-dim): $h | ref, \\gamma \\sim Categorical(W) $ \n",
    "with $W(l,i)=\n",
    "\\begin{cases} \n",
    "\\gamma,  \\text{ if }i = ref[l] \\\\\n",
    "\\frac{1-\\gamma}{4}, \\text{ else. }\n",
    "\\end{cases}$ for $l \\in {1, ..., L}$ and $i\\in {1,..., |alphabet|}$  \n",
    "Likelihood of the reads ($N$-dim): $r | z, h, \\theta \\sim Categorical(E)$ \n",
    "with $E(n,l,i)=\n",
    "\\begin{cases} \n",
    "\\theta,  \\text{ if }i = h_{z_n}[l] \\\\\n",
    "\\frac{1-\\theta}{4}, \\text{ else. }\n",
    "\\end{cases}$ for $n \\in {1, ..., N}$, $l \\in {1, ..., L}$ and $i\\in {1,..., |alphabet|}$  \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "id": "77ae49d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def model_finiteDPM(input_data):\n",
    "    reference, read_data, alphabet_length, cluster_num = input_data\n",
    "\n",
    "    # parameters\n",
    "    read_count = read_data.shape[0]\n",
    "    genome_length = read_data.shape[1]\n",
    "    alphabet_length = alphabet_length\n",
    "\n",
    "    # define rates\n",
    "    mutation_rate = numpyro.sample('mutation_rate', dist.Beta(1, 1))\n",
    "    error_rate = numpyro.sample('error_rate', dist.Beta(1, 1))\n",
    "\n",
    "    # create matrix of rates\n",
    "    mutation_rate_matrix = jnp.full(\n",
    "        (genome_length, alphabet_length), (1 - mutation_rate) / (alphabet_length - 1)\n",
    "    )\n",
    "    mutation_rate_matrix = custom_put_along_axis(\n",
    "        mutation_rate_matrix, reference.reshape(genome_length, 1), mutation_rate, axis=1\n",
    "    )\n",
    "\n",
    "    alpha0 = 0.001 # alpha must be bigger than zero\n",
    "    alpha = alpha0 * jnp.ones(cluster_num) / cluster_num\n",
    "    cluster_weights = numpyro.sample(\"cluster_weights\", dist.Dirichlet(jnp.ones(cluster_num)))\n",
    "\n",
    "    genome_axis = numpyro.plate(\"genome_axis\", genome_length, dim=-1)\n",
    "    with numpyro.plate(\"haplotype_axis\", cluster_num, dim=-2):\n",
    "        \n",
    "        with genome_axis:\n",
    "            haplotypes = numpyro.sample(\"haplotypes\", dist.Categorical(mutation_rate_matrix))  # cluster centers\n",
    "\n",
    "    with numpyro.plate(\"read_axis\", read_count, dim=-2):\n",
    "        cluster_assignments = numpyro.sample(\"cluster_assignments\", dist.Categorical(cluster_weights))\n",
    "        \n",
    "        with genome_axis:\n",
    "            \n",
    "            error_rate_matrix = jnp.full(\n",
    "                (read_count, genome_length, alphabet_length), (1 - error_rate) / (alphabet_length - 1)\n",
    "            )\n",
    "            error_rate_matrix = custom_put_along_axis(\n",
    "                error_rate_matrix, haplotypes[cluster_assignments].reshape(read_count, genome_length, 1), error_rate, axis=2\n",
    "            )\n",
    "\n",
    "            obs=numpyro.sample(\"obs\", dist.Categorical(error_rate_matrix), obs=read_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "72b832d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.50.0 (20211204.2007)\n",
       " -->\n",
       "<!-- Pages: 1 -->\n",
       "<svg width=\"442pt\" height=\"269pt\"\n",
       " viewBox=\"0.00 0.00 442.00 269.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 265)\">\n",
       "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-265 438,-265 438,4 -4,4\"/>\n",
       "<g id=\"clust1\" class=\"cluster\">\n",
       "<title>cluster_genome_axis</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"8,-8 8,-205 138,-205 138,-8 8,-8\"/>\n",
       "<text text-anchor=\"middle\" x=\"93.5\" y=\"-15.8\" font-family=\"Times,serif\" font-size=\"14.00\">genome_axis</text>\n",
       "</g>\n",
       "<g id=\"clust2\" class=\"cluster\">\n",
       "<title>cluster_haplotype_axis</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"16,-122 16,-197 130,-197 130,-122 16,-122\"/>\n",
       "<text text-anchor=\"middle\" x=\"81\" y=\"-129.8\" font-family=\"Times,serif\" font-size=\"14.00\">haplotype_axis</text>\n",
       "</g>\n",
       "<g id=\"clust3\" class=\"cluster\">\n",
       "<title>cluster_read_axis</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"60,-39 60,-114 130,-114 130,-39 60,-39\"/>\n",
       "<text text-anchor=\"middle\" x=\"96\" y=\"-46.8\" font-family=\"Times,serif\" font-size=\"14.00\">read_axis</text>\n",
       "</g>\n",
       "<g id=\"clust4\" class=\"cluster\">\n",
       "<title>cluster_read_axis__CLONE</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"246,-122 246,-197 426,-197 426,-122 246,-122\"/>\n",
       "<text text-anchor=\"middle\" x=\"392\" y=\"-129.8\" font-family=\"Times,serif\" font-size=\"14.00\">read_axis</text>\n",
       "</g>\n",
       "<!-- mutation_rate -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>mutation_rate</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"73\" cy=\"-243\" rx=\"59.59\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"73\" y=\"-239.3\" font-family=\"Times,serif\" font-size=\"14.00\">mutation_rate</text>\n",
       "</g>\n",
       "<!-- haplotypes -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>haplotypes</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"73\" cy=\"-171\" rx=\"48.99\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"73\" y=\"-167.3\" font-family=\"Times,serif\" font-size=\"14.00\">haplotypes</text>\n",
       "</g>\n",
       "<!-- mutation_rate&#45;&gt;haplotypes -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>mutation_rate&#45;&gt;haplotypes</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M73,-224.7C73,-216.98 73,-207.71 73,-199.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"76.5,-199.1 73,-189.1 69.5,-199.1 76.5,-199.1\"/>\n",
       "</g>\n",
       "<!-- error_rate -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>error_rate</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"191\" cy=\"-171\" rx=\"45.49\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"191\" y=\"-167.3\" font-family=\"Times,serif\" font-size=\"14.00\">error_rate</text>\n",
       "</g>\n",
       "<!-- obs -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>obs</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"95\" cy=\"-88\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"95\" y=\"-84.3\" font-family=\"Times,serif\" font-size=\"14.00\">obs</text>\n",
       "</g>\n",
       "<!-- error_rate&#45;&gt;obs -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>error_rate&#45;&gt;obs</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M172.49,-154.38C157.23,-141.5 135.47,-123.14 118.89,-109.16\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"121.12,-106.46 111.22,-102.68 116.6,-111.81 121.12,-106.46\"/>\n",
       "</g>\n",
       "<!-- cluster_weights -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>cluster_weights</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"336\" cy=\"-243\" rx=\"65.79\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"336\" y=\"-239.3\" font-family=\"Times,serif\" font-size=\"14.00\">cluster_weights</text>\n",
       "</g>\n",
       "<!-- cluster_assignments -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>cluster_assignments</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"336\" cy=\"-171\" rx=\"81.79\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"336\" y=\"-167.3\" font-family=\"Times,serif\" font-size=\"14.00\">cluster_assignments</text>\n",
       "</g>\n",
       "<!-- cluster_weights&#45;&gt;cluster_assignments -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>cluster_weights&#45;&gt;cluster_assignments</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M336,-224.7C336,-216.98 336,-207.71 336,-199.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"339.5,-199.1 336,-189.1 332.5,-199.1 339.5,-199.1\"/>\n",
       "</g>\n",
       "<!-- haplotypes&#45;&gt;obs -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>haplotypes&#45;&gt;obs</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M77.67,-152.82C80.6,-142.01 84.45,-127.84 87.78,-115.59\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"91.17,-116.44 90.42,-105.88 84.42,-114.61 91.17,-116.44\"/>\n",
       "</g>\n",
       "<!-- cluster_assignments&#45;&gt;obs -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>cluster_assignments&#45;&gt;obs</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M309.8,-153.75C292.07,-143.38 267.81,-130.35 245,-122 207.32,-108.2 162.11,-99.22 131.24,-94.17\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"131.77,-90.71 121.35,-92.6 130.68,-97.62 131.77,-90.71\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x7ffc901020a0>"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "numpyro.render_model(model_finiteDPM, model_args=(input_data,))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06f54e72",
   "metadata": {},
   "source": [
    "-----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4aa5df1",
   "metadata": {},
   "source": [
    "## Model - Infinite DPM - Stickbreaking process\n",
    "#### Dirichlet mixture model where number of clusters is learn.\n",
    "\n",
    "ref = reference sequence  \n",
    "$N$ = number of reads  \n",
    "$K$ = number of clusters/components  \n",
    "$L$ = genome length (number of positions)  \n",
    "alphabet = {A, C, G, T, -}\n",
    "\n",
    "\n",
    "no-mutation rate: $\\gamma \\sim Beta(a,b)$   \n",
    "no-error rate: $\\theta \\sim Beta(c,d)$  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "2f55b45d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mix_weights(beta):\n",
    "    beta1m_cumprod = (1 - beta).cumprod(-1)\n",
    "    return jnp.pad(beta, (0, 1), constant_values=1) * jnp.pad(\n",
    "        beta1m_cumprod, (1, 0), constant_values=1\n",
    "    )\n",
    "\n",
    "\n",
    "def model_infiniteSBP(input_data):\n",
    "    reference, read_data, alphabet_length, max_cluster_num_guess = input_data\n",
    "\n",
    "    max_cluster_num_guess = 10  # guess for maximum number of clusters\n",
    "\n",
    "    # parameters\n",
    "    read_count = read_data.shape[0]\n",
    "    genome_length = reference.shape[0]\n",
    "    alphabet_length = alphabet_length #jnp.unique(reference).shape[0]\n",
    "\n",
    "    # define rates\n",
    "    mutation_rate = numpyro.sample(\"mutation_rate\", dist.Beta(1, 1))\n",
    "    error_rate = numpyro.sample(\"error_rate\", dist.Beta(1, 1))\n",
    "\n",
    "    # create matrix of rates\n",
    "    mutation_rate_matrix = jnp.full(\n",
    "        (genome_length, alphabet_length), (1 - mutation_rate) / (alphabet_length - 1)\n",
    "    )\n",
    "    mutation_rate_matrix = custom_put_along_axis(\n",
    "        mutation_rate_matrix, reference.reshape(genome_length, 1), mutation_rate, axis=1\n",
    "    )\n",
    "\n",
    "    with numpyro.plate(\"beta_plate\", max_cluster_num_guess - 1):\n",
    "        beta = numpyro.sample(\"beta\", dist.Beta(1, 1.1))\n",
    "    \n",
    "    cluster_weights = numpyro.sample(\n",
    "        \"cluster_weights\", dist.Dirichlet(mix_weights(beta))\n",
    "    )\n",
    "\n",
    "    genome_axis = numpyro.plate(\"genome_axis\", genome_length, dim=-1)\n",
    "    with numpyro.plate(\"haplotype_axis\", max_cluster_num_guess, dim=-2):\n",
    "        with genome_axis:\n",
    "            haplotypes = numpyro.sample(\n",
    "                \"haplotypes\", dist.Categorical(mutation_rate_matrix)\n",
    "            )  # cluster centers\n",
    "\n",
    "    with numpyro.plate(\"read_axis\", read_count, dim=-2):\n",
    "        cluster_assignments = numpyro.sample(\n",
    "            \"cluster_assignments\", dist.Categorical(cluster_weights)\n",
    "        )\n",
    "\n",
    "        with genome_axis:\n",
    "            error_rate_matrix = jnp.full(\n",
    "                (read_count, genome_length, alphabet_length),\n",
    "                (1 - error_rate) / (alphabet_length - 1),\n",
    "            )\n",
    "            error_rate_matrix = custom_put_along_axis(\n",
    "                error_rate_matrix,\n",
    "                haplotypes[cluster_assignments].reshape(read_count, genome_length, 1),\n",
    "                error_rate,\n",
    "                axis=2,\n",
    "            )\n",
    "\n",
    "            numpyro.sample(\"obs\", dist.Categorical(error_rate_matrix), obs=read_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "3cf28e8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.50.0 (20211204.2007)\n",
       " -->\n",
       "<!-- Pages: 1 -->\n",
       "<svg width=\"442pt\" height=\"360pt\"\n",
       " viewBox=\"0.00 0.00 442.00 360.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 356)\">\n",
       "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-356 438,-356 438,4 -4,4\"/>\n",
       "<g id=\"clust1\" class=\"cluster\">\n",
       "<title>cluster_beta_plate</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"300,-269 300,-344 372,-344 372,-269 300,-269\"/>\n",
       "<text text-anchor=\"middle\" x=\"336\" y=\"-276.8\" font-family=\"Times,serif\" font-size=\"14.00\">beta_plate</text>\n",
       "</g>\n",
       "<g id=\"clust2\" class=\"cluster\">\n",
       "<title>cluster_genome_axis</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"8,-8 8,-205 138,-205 138,-8 8,-8\"/>\n",
       "<text text-anchor=\"middle\" x=\"93.5\" y=\"-15.8\" font-family=\"Times,serif\" font-size=\"14.00\">genome_axis</text>\n",
       "</g>\n",
       "<g id=\"clust3\" class=\"cluster\">\n",
       "<title>cluster_haplotype_axis</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"16,-122 16,-197 130,-197 130,-122 16,-122\"/>\n",
       "<text text-anchor=\"middle\" x=\"81\" y=\"-129.8\" font-family=\"Times,serif\" font-size=\"14.00\">haplotype_axis</text>\n",
       "</g>\n",
       "<g id=\"clust4\" class=\"cluster\">\n",
       "<title>cluster_read_axis</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"60,-39 60,-114 130,-114 130,-39 60,-39\"/>\n",
       "<text text-anchor=\"middle\" x=\"96\" y=\"-46.8\" font-family=\"Times,serif\" font-size=\"14.00\">read_axis</text>\n",
       "</g>\n",
       "<g id=\"clust5\" class=\"cluster\">\n",
       "<title>cluster_read_axis__CLONE</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"246,-122 246,-197 426,-197 426,-122 246,-122\"/>\n",
       "<text text-anchor=\"middle\" x=\"392\" y=\"-129.8\" font-family=\"Times,serif\" font-size=\"14.00\">read_axis</text>\n",
       "</g>\n",
       "<!-- mutation_rate -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>mutation_rate</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"73\" cy=\"-243\" rx=\"59.59\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"73\" y=\"-239.3\" font-family=\"Times,serif\" font-size=\"14.00\">mutation_rate</text>\n",
       "</g>\n",
       "<!-- haplotypes -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>haplotypes</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"73\" cy=\"-171\" rx=\"48.99\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"73\" y=\"-167.3\" font-family=\"Times,serif\" font-size=\"14.00\">haplotypes</text>\n",
       "</g>\n",
       "<!-- mutation_rate&#45;&gt;haplotypes -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>mutation_rate&#45;&gt;haplotypes</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M73,-224.7C73,-216.98 73,-207.71 73,-199.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"76.5,-199.1 73,-189.1 69.5,-199.1 76.5,-199.1\"/>\n",
       "</g>\n",
       "<!-- error_rate -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>error_rate</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"191\" cy=\"-171\" rx=\"45.49\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"191\" y=\"-167.3\" font-family=\"Times,serif\" font-size=\"14.00\">error_rate</text>\n",
       "</g>\n",
       "<!-- obs -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>obs</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"95\" cy=\"-88\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"95\" y=\"-84.3\" font-family=\"Times,serif\" font-size=\"14.00\">obs</text>\n",
       "</g>\n",
       "<!-- error_rate&#45;&gt;obs -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>error_rate&#45;&gt;obs</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M172.49,-154.38C157.23,-141.5 135.47,-123.14 118.89,-109.16\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"121.12,-106.46 111.22,-102.68 116.6,-111.81 121.12,-106.46\"/>\n",
       "</g>\n",
       "<!-- cluster_weights -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>cluster_weights</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"336\" cy=\"-243\" rx=\"65.79\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"336\" y=\"-239.3\" font-family=\"Times,serif\" font-size=\"14.00\">cluster_weights</text>\n",
       "</g>\n",
       "<!-- cluster_assignments -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>cluster_assignments</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"336\" cy=\"-171\" rx=\"81.79\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"336\" y=\"-167.3\" font-family=\"Times,serif\" font-size=\"14.00\">cluster_assignments</text>\n",
       "</g>\n",
       "<!-- cluster_weights&#45;&gt;cluster_assignments -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>cluster_weights&#45;&gt;cluster_assignments</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M336,-224.7C336,-216.98 336,-207.71 336,-199.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"339.5,-199.1 336,-189.1 332.5,-199.1 339.5,-199.1\"/>\n",
       "</g>\n",
       "<!-- beta -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>beta</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"336\" cy=\"-318\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"336\" y=\"-314.3\" font-family=\"Times,serif\" font-size=\"14.00\">beta</text>\n",
       "</g>\n",
       "<!-- beta&#45;&gt;cluster_weights -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>beta&#45;&gt;cluster_weights</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M336,-299.7C336,-291.25 336,-280.87 336,-271.37\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"339.5,-271.18 336,-261.18 332.5,-271.18 339.5,-271.18\"/>\n",
       "</g>\n",
       "<!-- haplotypes&#45;&gt;obs -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>haplotypes&#45;&gt;obs</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M77.67,-152.82C80.6,-142.01 84.45,-127.84 87.78,-115.59\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"91.17,-116.44 90.42,-105.88 84.42,-114.61 91.17,-116.44\"/>\n",
       "</g>\n",
       "<!-- cluster_assignments&#45;&gt;obs -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>cluster_assignments&#45;&gt;obs</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M309.8,-153.75C292.07,-143.38 267.81,-130.35 245,-122 207.32,-108.2 162.11,-99.22 131.24,-94.17\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"131.77,-90.71 121.35,-92.6 130.68,-97.62 131.77,-90.71\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x7ffc8e9f43a0>"
      ]
     },
     "execution_count": 222,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "numpyro.render_model(model_infiniteSBP, model_args=(input_data,))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c78d787c",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56ca44c4",
   "metadata": {},
   "source": [
    "## Model - Finite DPM\n",
    "\n",
    "#### Finite Dirichlet mixture model with fixed number of clusters. With extension of error and mutation model accounting for number of matches, number of subsitutions, number of deletions\n",
    "\n",
    "ref = reference sequence  \n",
    "$N$ = number of reads  \n",
    "$K$ = number of clusters/components  \n",
    "$L$ = genome length (number of positions)  \n",
    "alphabet = {A, C, G, T, -}\n",
    "\n",
    "\n",
    "mutation rate: $\\gamma, \\gamma_S, \\gamma_{D^f} \\gamma_{D^b}\\sim Dirichlet(a,b,c,d)$  \n",
    "\n",
    "mutation rate matrix depending on the reference:\n",
    "$\n",
    "\\begin{matrix}\n",
    "ref & A & C & G & T & - \\\\\n",
    "A & \\gamma & \\gamma_S & \\gamma_S & \\gamma_S & \\gamma_{D^b} \\\\\n",
    "C & \\gamma_S & \\gamma & \\gamma_S & \\gamma_S & \\gamma_{D^b} \\\\\n",
    "G & \\gamma_S & \\gamma_S & \\gamma & \\gamma_S & \\gamma_{D^b} \\\\\n",
    "T & \\gamma_S & \\gamma_S & \\gamma_S & \\gamma & \\gamma_{D^b} \\\\\n",
    "- & \\gamma_{D^f} & \\gamma_{D^f}& \\gamma_{D^f} & \\gamma_{D^f} & \\gamma \\\\\n",
    "\\end{matrix}\n",
    "$\n",
    "\n",
    "error rate: $\\theta, \\theta_S, \\theta_{D^f} \\theta_{D^b} \\sim Dirichlet(a,b,c,d)$   \n",
    "\n",
    "error rate matrix depending on the assigned haplotype:\n",
    "$\n",
    "\\begin{matrix}\n",
    "haplotype & A & C & G & T & - \\\\\n",
    "A & \\theta & \\theta_S & \\theta_S & \\theta_S & \\theta_{D^b} \\\\\n",
    "C & \\theta_S & \\theta & \\theta_S & \\theta_S & \\theta_{D^b} \\\\\n",
    "G & \\theta_S & \\theta_S & \\theta & \\theta_S & \\theta_{D^b} \\\\\n",
    "T & \\theta_S & \\theta_S & \\theta_S & \\theta & \\theta_{D^b} \\\\\n",
    "- & \\theta_{D^f} & \\theta_{D^f}& \\theta_{D^f} & \\theta_{D^f} & \\theta \\\\\n",
    "\\end{matrix}\n",
    "$\n",
    "\n",
    "\n",
    "Cluster weights ($K$-dim): $\\pi | \\alpha \\sim Dir(\\alpha)$ \n",
    "\n",
    "Cluster assignments ($N$-dim): $z|\\pi \\sim Categorical(\\pi)$  \n",
    "\n",
    "Cluster centers/haplotypes ($K$x$L$-dim): $h | ref, \\gamma \\sim Categorical(W) $ \n",
    "with $W(l,i)=$ mutation rates (see above).\n",
    "\n",
    "Likelihood of the reads ($N$-dim): $r | z, h, \\theta \\sim Categorical(E)$ \n",
    "with $E(n,l,i)=$  error rate matrix (see above).\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "19d99336",
   "metadata": {},
   "outputs": [],
   "source": [
    "def model_finiteDPM_extended(input_data):\n",
    "    '''\n",
    "    This model is only for alphabet={A,C,G,T,-} as deletion mutations and errors are treated seperatly \n",
    "    (wrt to subsitutions).\n",
    "    '''\n",
    "    cluster_num = 10\n",
    "    reference, read_data, alphabet_length = input_data\n",
    "\n",
    "    # parameters\n",
    "    read_count = read_data.shape[0]\n",
    "    genome_length = read_data.shape[1]\n",
    "    alphabet_length = alphabet_length\n",
    "\n",
    "    # define rates\n",
    "    mutation_rate = numpyro.sample('mutation_rate', dist.Dirichlet(jnp.ones(4)/4))\n",
    "    error_rate = numpyro.sample('error_rate', dist.Dirichlet(jnp.ones(4)/4))\n",
    "\n",
    "    # create matrix of rates\n",
    "    mutation_rate_matrix = jnp.full(\n",
    "        (genome_length, alphabet_length), mutation_rate[1] / (alphabet_length - 2)\n",
    "    ) # subsitutions\n",
    "    mutation_rate_matrix = mutation_rate_matrix.at[:,4].set(mutation_rate[2]) # deletion in haplotype\n",
    "    \n",
    "    '''\n",
    "    # TODO: \n",
    "    # Option 0: \n",
    "    mutation_rate_matrix = mutation_rate_matrix.at[indicies,:].set(mutation_rate[3] / (alphabet_length - 1))\n",
    "    # Option 1: \n",
    "    mutation_rate_matrix = mutation_rate_matrix.at[indicies,:].set((mutation_rate[3] / (alphabet_length - 1))*jnp.ones(5))\n",
    "    # Option 2: \n",
    "    indicies = jnp.argwhere(reference==4)\n",
    "    mutation_rate_matrix = mutation_rate_matrix.at[indicies,0].set(\n",
    "        mutation_rate[3] / (alphabet_length - 1)\n",
    "    )\n",
    "    mutation_rate_matrix = mutation_rate_matrix.at[indicies,1].set(\n",
    "        mutation_rate[3] / (alphabet_length - 1)\n",
    "    )\n",
    "    mutation_rate_matrix = mutation_rate_matrix.at[indicies,2].set(\n",
    "        mutation_rate[3] / (alphabet_length - 1)\n",
    "    )\n",
    "    mutation_rate_matrix = mutation_rate_matrix.at[indicies,3].set(\n",
    "        mutation_rate[3] / (alphabet_length - 1)\n",
    "    )\n",
    "    mutation_rate_matrix = mutation_rate_matrix.at[indicies,4].set(\n",
    "        mutation_rate[0]\n",
    "    )\n",
    "    '''\n",
    "    mutation_rate_matrix = custom_put_along_axis(\n",
    "        mutation_rate_matrix, reference.reshape(genome_length, 1), mutation_rate[0], axis=1\n",
    "    ) # matches\n",
    "    mutation_rate_matrix = mutation_rate_matrix / mutation_rate_matrix.sum(axis=1)[:,jnp.newaxis] #normalize\n",
    "    \n",
    "\n",
    "    alpha0 = 0.001 # alpha must be bigger than zero\n",
    "    alpha = alpha0 * jnp.ones(cluster_num) / cluster_num\n",
    "    cluster_weights = numpyro.sample(\"cluster_weights\", dist.Dirichlet(jnp.ones(cluster_num)))\n",
    "\n",
    "    genome_axis = numpyro.plate(\"genome_axis\", genome_length, dim=-1)\n",
    "    with numpyro.plate(\"haplotype_axis\", cluster_num, dim=-2):\n",
    "        \n",
    "        with genome_axis:\n",
    "            haplotypes = numpyro.sample(\"haplotypes\", dist.Categorical(mutation_rate_matrix))  # cluster centers\n",
    "    \n",
    "    with numpyro.plate(\"read_axis\", read_count, dim=-2):\n",
    "        cluster_assignments = numpyro.sample(\"cluster_assignments\", dist.Categorical(cluster_weights))\n",
    "        with genome_axis:\n",
    "            error_rate_matrix = jnp.full(\n",
    "                (read_count, genome_length, alphabet_length),\n",
    "                error_rate[1] / (alphabet_length - 1),\n",
    "            ) # substitutions \n",
    "            error_rate_matrix = error_rate_matrix.at[:,:,4].set(error_rate[2]) # deletion in read\n",
    "            '''\n",
    "            # TODO: Find solution to this problem!! It is doing the right thing but its messing up \n",
    "            # after one or two iterations\n",
    "            # deletion in haplotype:\n",
    "            # this are the indices where we want to have theta_Db\n",
    "            ind_list = jnp.argwhere(haplotypes[cluster_assignments].reshape(read_count, genome_length)==4)\n",
    "            for ind in ind_list:\n",
    "                error_rate_matrix = error_rate_matrix.at[ind[0],ind[1]].set(error_rate[3]) # deletion in haplotype \n",
    "            '''\n",
    "            error_rate_matrix = custom_put_along_axis(\n",
    "                error_rate_matrix,\n",
    "                haplotypes[cluster_assignments].reshape(read_count, genome_length, 1),\n",
    "                error_rate[0],\n",
    "                axis=2,\n",
    "            ) # matches \n",
    "\n",
    "            numpyro.sample(\"obs\", dist.Categorical(error_rate_matrix), obs=read_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "be2cc9b1",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "too many values to unpack (expected 3)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/dw/8dl4p6h53cgcmk6cf_09v5f40000gr/T/ipykernel_4528/139958247.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mnumpyro\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrender_model\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodel_finiteDPM_extended\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmodel_args\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput_data\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/numpyro/contrib/render.py\u001b[0m in \u001b[0;36mrender_model\u001b[0;34m(model, model_args, model_kwargs, filename, render_distributions)\u001b[0m\n\u001b[1;32m    297\u001b[0m     \u001b[0;34m:\u001b[0m\u001b[0mparam\u001b[0m \u001b[0mbool\u001b[0m \u001b[0mrender_distributions\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mWhether\u001b[0m \u001b[0mto\u001b[0m \u001b[0minclude\u001b[0m \u001b[0mRV\u001b[0m \u001b[0mdistribution\u001b[0m \u001b[0mannotations\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mplot\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    298\u001b[0m     \"\"\"\n\u001b[0;32m--> 299\u001b[0;31m     relations = get_model_relations(\n\u001b[0m\u001b[1;32m    300\u001b[0m         \u001b[0mmodel\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    301\u001b[0m         \u001b[0mmodel_args\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmodel_args\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/numpyro/contrib/render.py\u001b[0m in \u001b[0;36mget_model_relations\u001b[0;34m(model, model_args, model_kwargs)\u001b[0m\n\u001b[1;32m     76\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     77\u001b[0m     \u001b[0;31m# We use eval_shape to avoid any array computation.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 78\u001b[0;31m     \u001b[0mtrace\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mjax\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0meval_shape\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mget_trace\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtrace\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     79\u001b[0m     obs_sites = [\n\u001b[1;32m     80\u001b[0m         \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/jax/_src/api.py\u001b[0m in \u001b[0;36meval_shape\u001b[0;34m(fun, *args, **kwargs)\u001b[0m\n\u001b[1;32m   2952\u001b[0m   \u001b[0mwrapped_fun\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout_tree\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mflatten_fun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlu\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrap_init\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfun\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0min_tree\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2953\u001b[0m   \u001b[0mdebug_info\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpe\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdebug_info\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfun\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0min_tree\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"eval_shape\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2954\u001b[0;31m   out = pe.abstract_eval_fun(wrapped_fun.call_wrapped,\n\u001b[0m\u001b[1;32m   2955\u001b[0m                              \u001b[0;34m*\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mshaped_abstractify\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs_flat\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2956\u001b[0m                              debug_info=debug_info)\n",
      "\u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/jax/interpreters/partial_eval.py\u001b[0m in \u001b[0;36mabstract_eval_fun\u001b[0;34m(fun, debug_info, *avals, **params)\u001b[0m\n\u001b[1;32m    417\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    418\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mabstract_eval_fun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfun\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0mavals\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdebug_info\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mparams\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 419\u001b[0;31m   _, avals_out, _ = trace_to_jaxpr_dynamic(\n\u001b[0m\u001b[1;32m    420\u001b[0m       lu.wrap_init(fun, params), avals, debug_info)\n\u001b[1;32m    421\u001b[0m   \u001b[0;32massert\u001b[0m \u001b[0mall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maval\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mAbstractValue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0maval\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mavals_out\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/jax/_src/profiler.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    204\u001b[0m   \u001b[0;32mdef\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    205\u001b[0m     \u001b[0;32mwith\u001b[0m \u001b[0mTraceAnnotation\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mdecorator_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 206\u001b[0;31m       \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    207\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    208\u001b[0m   \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/jax/interpreters/partial_eval.py\u001b[0m in \u001b[0;36mtrace_to_jaxpr_dynamic\u001b[0;34m(fun, in_avals, debug_info)\u001b[0m\n\u001b[1;32m   1531\u001b[0m     \u001b[0mmain\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdebug_info\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdebug_info\u001b[0m  \u001b[0;31m# type: ignore\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1532\u001b[0m     \u001b[0mmain\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjaxpr_stack\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# type: ignore\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1533\u001b[0;31m     \u001b[0mjaxpr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout_avals\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconsts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtrace_to_subjaxpr_dynamic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfun\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmain\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0min_avals\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1534\u001b[0m     \u001b[0;32mdel\u001b[0m \u001b[0mmain\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfun\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1535\u001b[0m   \u001b[0;32mreturn\u001b[0m \u001b[0mjaxpr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout_avals\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconsts\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/jax/interpreters/partial_eval.py\u001b[0m in \u001b[0;36mtrace_to_subjaxpr_dynamic\u001b[0;34m(fun, main, in_avals)\u001b[0m\n\u001b[1;32m   1541\u001b[0m     \u001b[0mtrace\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mDynamicJaxprTrace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmain\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcore\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcur_sublevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1542\u001b[0m     \u001b[0min_tracers\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtrace\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnew_arg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0min_avals\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1543\u001b[0;31m     \u001b[0mans\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfun\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcall_wrapped\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0min_tracers\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1544\u001b[0m     \u001b[0mout_tracers\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtrace\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfull_raise\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mans\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1545\u001b[0m     \u001b[0mjaxpr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout_avals\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconsts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mframe\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_jaxpr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0min_tracers\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout_tracers\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/jax/linear_util.py\u001b[0m in \u001b[0;36mcall_wrapped\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    164\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    165\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 166\u001b[0;31m       \u001b[0mans\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mdict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparams\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    167\u001b[0m     \u001b[0;32mexcept\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    168\u001b[0m       \u001b[0;31m# Some transformations yield from inside context managers, so we have to\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/jax/linear_util.py\u001b[0m in \u001b[0;36mcall_wrapped\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    164\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    165\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 166\u001b[0;31m       \u001b[0mans\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mdict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparams\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    167\u001b[0m     \u001b[0;32mexcept\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    168\u001b[0m       \u001b[0;31m# Some transformations yield from inside context managers, so we have to\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/numpyro/contrib/render.py\u001b[0m in \u001b[0;36mget_trace\u001b[0;34m()\u001b[0m\n\u001b[1;32m     66\u001b[0m             \u001b[0msubstitute_fn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minit_to_sample\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     67\u001b[0m         )\n\u001b[0;32m---> 68\u001b[0;31m         \u001b[0mtrace\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhandlers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtrace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubs_model\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_trace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mmodel_args\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mmodel_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     69\u001b[0m         \u001b[0;31m# Work around an issue where jax.eval_shape does not work\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     70\u001b[0m         \u001b[0;31m# for distribution output (e.g. the function `lambda: dist.Normal(0, 1)`)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/numpyro/handlers.py\u001b[0m in \u001b[0;36mget_trace\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    169\u001b[0m         \u001b[0;34m:\u001b[0m\u001b[0;32mreturn\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m`\u001b[0m\u001b[0mOrderedDict\u001b[0m\u001b[0;31m`\u001b[0m \u001b[0mcontaining\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mexecution\u001b[0m \u001b[0mtrace\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    170\u001b[0m         \"\"\"\n\u001b[0;32m--> 171\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    172\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtrace\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    173\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/numpyro/primitives.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m     85\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     86\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 87\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     88\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     89\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/numpyro/primitives.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m     85\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     86\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 87\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     88\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     89\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/numpyro/primitives.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m     85\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     86\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 87\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     88\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     89\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/folders/dw/8dl4p6h53cgcmk6cf_09v5f40000gr/T/ipykernel_4528/3318714468.py\u001b[0m in \u001b[0;36mmodel_finiteDPM_extended\u001b[0;34m(input_data)\u001b[0m\n\u001b[1;32m      5\u001b[0m     '''\n\u001b[1;32m      6\u001b[0m     \u001b[0mcluster_num\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m10\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m     \u001b[0mreference\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mread_data\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0malphabet_length\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minput_data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m     \u001b[0;31m# parameters\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: too many values to unpack (expected 3)"
     ]
    }
   ],
   "source": [
    "numpyro.render_model(model_finiteDPM_extended, model_args=(input_data,))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35140a87",
   "metadata": {},
   "source": [
    "## Model - Hierarchical Dirichlet Process Mixture (hierDPM)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6dc30e07",
   "metadata": {},
   "source": [
    "$m \\in \\{1, ..., M\\}$ samples denoted as $\\{x_1, ..., x_M \\}$ where sample $x_m = \\{r_1, ..., r_{n_m}\\}$  \n",
    "Note that every sample can have a different amount of reads.  \n",
    "$k \\in \\{1, ..., K\\}$ mixtures/clusters\n",
    "\n",
    "For now lets have $\\gamma$ (mutation rate) and $\\theta$ (error rate) fixed. \n",
    "\n",
    "$p(x_1, ... , x_M) = \\prod _{m=1}^M p()$\n",
    "\n",
    "where:\n",
    "\n",
    "$z_{mn} \\sim Categorical(\\pi_m)$ and $\\pi_m =(\\pi_{m1}, ...,\\pi_{mK})$\n",
    "\n",
    "for each $m \\in \\{1, ..., M\\}$: $\\pi_m \\sim Dirichlet(\\alpha)$ with $\\alpha$ being $K$-dimensional\n",
    "\n",
    "haplotypes ($K$x$L$-dim): $h | ref, \\gamma \\sim Categorical(M)$ with $M$ mutation rate matrix (similar defined like above)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "bcbfff19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Minimal example input\n",
    "\n",
    "reference = jnp.array([0])\n",
    "reads = jnp.array([[[0], [1], [1], [1], [0], [1], [0], [1]], [[0], [0], [0], [1], [0], [1], [0], [1]]])\n",
    "alphabet ='01'\n",
    "\n",
    "cluster_num = 5\n",
    "\n",
    "input_data = reference, reads, len(alphabet), cluster_num"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cf596dd",
   "metadata": {},
   "source": [
    "### TODO: How to solve the problem that different samples will have a different amount of reads?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "39ec401b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mix_weights(beta):\n",
    "    beta1m_cumprod = (1 - beta).cumprod(-1)\n",
    "    pad_beta = jnp.pad(beta, ((0, 0),(0, 1)), constant_values=1)\n",
    "    pad_beta1m_cumprod = jnp.pad(beta1m_cumprod, ((0, 0),(1, 0)), constant_values=1)\n",
    "    \n",
    "    return pad_beta*pad_beta1m_cumprod\n",
    "\n",
    "def shape_parameters(alpha, beta):\n",
    "    # first shape parameter of beta distribution \n",
    "    first = alpha * beta\n",
    "    \n",
    "    # second shape parameter of beta distribution \n",
    "    beta_cumsum = beta.cumsum(-1) # K-1 dimensional \n",
    "    second = jnp.pad(beta_cumsum, (1, 0), constant_values=0)\n",
    "    second = second[:-1]\n",
    "    second = alpha * (1 - second)\n",
    "    \n",
    "    return [first, second]\n",
    "\n",
    "\n",
    "def model_hierarchicalSBMM(input_data):\n",
    "    reference, read_data, alphabet_length, max_cluster_num_guess = input_data\n",
    "\n",
    "    # parameters\n",
    "    samples_count = read_data.shape[0]\n",
    "    read_count = read_data.shape[1]\n",
    "    genome_length = reference.shape[0]\n",
    "    alphabet_length = alphabet_length \n",
    "    \n",
    "    # set hyperparameters\n",
    "    gamma = 1.1\n",
    "    alpha = 1.1\n",
    "    \n",
    "    # define rates\n",
    "    mutation_rate = numpyro.sample(\"mutation_rate\", dist.Beta(1, 1))\n",
    "    error_rate = numpyro.sample(\"error_rate\", dist.Beta(1, 1))\n",
    "\n",
    "    # create matrix of rates\n",
    "    mutation_rate_matrix = jnp.full(\n",
    "        (genome_length, alphabet_length), (1 - mutation_rate) / (alphabet_length - 1)\n",
    "    )\n",
    "    mutation_rate_matrix = custom_put_along_axis(\n",
    "        mutation_rate_matrix, reference.reshape(genome_length, 1), mutation_rate, axis=1\n",
    "    )\n",
    "    \n",
    "    # top hierarchical level - base distribution for all samples\n",
    "    beta_plate = numpyro.plate(\"beta_plate\", max_cluster_num_guess - 1)\n",
    "    with beta_plate:\n",
    "        beta_top = numpyro.sample(\"beta_top\", dist.Beta(1*jnp.ones(max_cluster_num_guess - 1), gamma*jnp.ones(max_cluster_num_guess - 1)))\n",
    "\n",
    "    # sample level - draw weights for each sample\n",
    "    sample_plate = numpyro.plate(\"sample_plate\", samples_count, dim=-2)\n",
    "    with sample_plate:\n",
    "        param = shape_parameters(alpha, beta_top)\n",
    "        beta = numpyro.sample(\"beta\", dist.Beta(param[0],param[1]))\n",
    "    \n",
    "    cluster_weights = numpyro.sample(\n",
    "    \"cluster_weights\", dist.Dirichlet(mix_weights(beta))\n",
    "    ) # n_samples X num_cluster\n",
    "        \n",
    "    genome_axis = numpyro.plate(\"genome_axis\", genome_length, dim=-1)\n",
    "    with numpyro.plate(\"haplotype_axis\", max_cluster_num_guess, dim=-2):\n",
    "        with genome_axis:\n",
    "            haplotypes = numpyro.sample(\n",
    "                \"haplotypes\", dist.Categorical(mutation_rate_matrix)\n",
    "            )  # cluster centers\n",
    "\n",
    "    with numpyro.plate(\"sample_plate\", samples_count, dim=-2):\n",
    "        with numpyro.plate(\"read_axis\", read_count, dim=-1):\n",
    "            cluster_weights = jnp.repeat(cluster_weights[:,jnp.newaxis,:], read_count, axis=1)\n",
    "            cluster_assignments = numpyro.sample(\n",
    "                \"cluster_assignments\", dist.Categorical(cluster_weights)\n",
    "            ) # n_samples x n_reads \n",
    "            \n",
    "    with numpyro.plate(\"sample_plate\", samples_count, dim=-3):    \n",
    "        with numpyro.plate(\"read_axis\", read_count, dim=-2):\n",
    "            with genome_axis:\n",
    "                error_rate_matrix = jnp.full(\n",
    "                    (samples_count, read_count, genome_length, alphabet_length),\n",
    "                    (1 - error_rate) / (alphabet_length - 1),\n",
    "                )\n",
    "                error_rate_matrix = custom_put_along_axis(\n",
    "                    error_rate_matrix,\n",
    "                    haplotypes[cluster_assignments].reshape(samples_count, read_count, genome_length, 1),\n",
    "                    error_rate,\n",
    "                    axis=3,\n",
    "                )\n",
    "\n",
    "                numpyro.sample(\"obs\", dist.Categorical(error_rate_matrix), obs=read_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "100ec244",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.50.0 (20211204.2007)\n",
       " -->\n",
       "<!-- Pages: 1 -->\n",
       "<svg width=\"514pt\" height=\"366pt\"\n",
       " viewBox=\"0.00 0.00 514.50 366.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 362)\">\n",
       "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-362 510.5,-362 510.5,4 -4,4\"/>\n",
       "<g id=\"clust1\" class=\"cluster\">\n",
       "<title>cluster_beta_plate</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"322,-275 322,-350 420,-350 420,-275 322,-275\"/>\n",
       "<text text-anchor=\"middle\" x=\"384\" y=\"-282.8\" font-family=\"Times,serif\" font-size=\"14.00\">beta_plate</text>\n",
       "</g>\n",
       "<g id=\"clust2\" class=\"cluster\">\n",
       "<title>cluster_sample_plate</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"146,-8 146,-267 406,-267 406,-8 146,-8\"/>\n",
       "<text text-anchor=\"middle\" x=\"361.5\" y=\"-15.8\" font-family=\"Times,serif\" font-size=\"14.00\">sample_plate</text>\n",
       "</g>\n",
       "<g id=\"clust3\" class=\"cluster\">\n",
       "<title>cluster_read_axis</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"154,-39 154,-259 334,-259 334,-39 154,-39\"/>\n",
       "<text text-anchor=\"middle\" x=\"300\" y=\"-46.8\" font-family=\"Times,serif\" font-size=\"14.00\">read_axis</text>\n",
       "</g>\n",
       "<g id=\"clust4\" class=\"cluster\">\n",
       "<title>cluster_genome_axis</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"200,-70 200,-145 289,-145 289,-70 200,-70\"/>\n",
       "<text text-anchor=\"middle\" x=\"244.5\" y=\"-77.8\" font-family=\"Times,serif\" font-size=\"14.00\">genome_axis</text>\n",
       "</g>\n",
       "<g id=\"clust5\" class=\"cluster\">\n",
       "<title>cluster_haplotype_axis</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"8,-153 8,-267 138,-267 138,-153 8,-153\"/>\n",
       "<text text-anchor=\"middle\" x=\"89\" y=\"-160.8\" font-family=\"Times,serif\" font-size=\"14.00\">haplotype_axis</text>\n",
       "</g>\n",
       "<g id=\"clust6\" class=\"cluster\">\n",
       "<title>cluster_genome_axis__CLONE</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"16,-184 16,-259 130,-259 130,-184 16,-184\"/>\n",
       "<text text-anchor=\"middle\" x=\"85.5\" y=\"-191.8\" font-family=\"Times,serif\" font-size=\"14.00\">genome_axis</text>\n",
       "</g>\n",
       "<!-- mutation_rate -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>mutation_rate</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"73\" cy=\"-324\" rx=\"59.59\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"73\" y=\"-320.3\" font-family=\"Times,serif\" font-size=\"14.00\">mutation_rate</text>\n",
       "</g>\n",
       "<!-- haplotypes -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>haplotypes</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"73\" cy=\"-233\" rx=\"48.99\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"73\" y=\"-229.3\" font-family=\"Times,serif\" font-size=\"14.00\">haplotypes</text>\n",
       "</g>\n",
       "<!-- mutation_rate&#45;&gt;haplotypes -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>mutation_rate&#45;&gt;haplotypes</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M73,-305.84C73,-293.28 73,-275.98 73,-261.5\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"76.5,-261.11 73,-251.11 69.5,-261.11 76.5,-261.11\"/>\n",
       "</g>\n",
       "<!-- error_rate -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>error_rate</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"461\" cy=\"-233\" rx=\"45.49\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"461\" y=\"-229.3\" font-family=\"Times,serif\" font-size=\"14.00\">error_rate</text>\n",
       "</g>\n",
       "<!-- obs -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>obs</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"244\" cy=\"-119\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"244\" y=\"-115.3\" font-family=\"Times,serif\" font-size=\"14.00\">obs</text>\n",
       "</g>\n",
       "<!-- error_rate&#45;&gt;obs -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>error_rate&#45;&gt;obs</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M454.69,-214.89C447.03,-196.58 432.28,-168.14 410,-153 371.51,-126.84 317.12,-120.42 281.22,-119.33\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"281.23,-115.83 271.17,-119.14 281.1,-122.82 281.23,-115.83\"/>\n",
       "</g>\n",
       "<!-- cluster_weights -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>cluster_weights</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"245\" cy=\"-324\" rx=\"65.79\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"245\" y=\"-320.3\" font-family=\"Times,serif\" font-size=\"14.00\">cluster_weights</text>\n",
       "</g>\n",
       "<!-- cluster_assignments -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>cluster_assignments</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"244\" cy=\"-233\" rx=\"81.79\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"244\" y=\"-229.3\" font-family=\"Times,serif\" font-size=\"14.00\">cluster_assignments</text>\n",
       "</g>\n",
       "<!-- cluster_weights&#45;&gt;cluster_assignments -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>cluster_weights&#45;&gt;cluster_assignments</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M244.81,-305.84C244.67,-293.28 244.47,-275.98 244.31,-261.5\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"247.8,-261.07 244.19,-251.11 240.8,-261.15 247.8,-261.07\"/>\n",
       "</g>\n",
       "<!-- beta_top -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>beta_top</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"371\" cy=\"-324\" rx=\"40.89\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"371\" y=\"-320.3\" font-family=\"Times,serif\" font-size=\"14.00\">beta_top</text>\n",
       "</g>\n",
       "<!-- beta -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>beta</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"371\" cy=\"-233\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"371\" y=\"-229.3\" font-family=\"Times,serif\" font-size=\"14.00\">beta</text>\n",
       "</g>\n",
       "<!-- beta_top&#45;&gt;beta -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>beta_top&#45;&gt;beta</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M371,-305.84C371,-293.28 371,-275.98 371,-261.5\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"374.5,-261.11 371,-251.11 367.5,-261.11 374.5,-261.11\"/>\n",
       "</g>\n",
       "<!-- beta&#45;&gt;cluster_weights -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>beta&#45;&gt;cluster_weights</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M358.3,-249.17C352.58,-255.31 345.44,-262.09 338,-267 330.01,-272.27 326.49,-270.57 318,-275 303.47,-282.58 288.2,-292.32 275.4,-301.04\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"273.21,-298.3 266.98,-306.87 277.19,-304.05 273.21,-298.3\"/>\n",
       "</g>\n",
       "<!-- cluster_assignments&#45;&gt;obs -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>cluster_assignments&#45;&gt;obs</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M244,-214.99C244,-197.06 244,-168.56 244,-147.44\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"247.5,-147.17 244,-137.17 240.5,-147.17 247.5,-147.17\"/>\n",
       "</g>\n",
       "<!-- haplotypes&#45;&gt;obs -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>haplotypes&#45;&gt;obs</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M85.43,-215.27C99.37,-197.56 123.52,-169.89 150,-153 167.96,-141.54 190.4,-133.47 208.85,-128.18\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"209.91,-131.52 218.64,-125.52 208.07,-124.76 209.91,-131.52\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x7ffc8ec150d0>"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "numpyro.render_model(model_hierarchicalSBMM, model_args=(input_data,))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9f2e335",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
