{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9540eedc",
   "metadata": {},
   "source": [
    "# Notebook with working models in NumPyro"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2778dd2b",
   "metadata": {},
   "source": [
    "--------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cf98a949",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load necessary packages\n",
    "import numpyro\n",
    "import numpyro.distributions as dist\n",
    "\n",
    "import jax\n",
    "import jax.numpy as jnp\n",
    "\n",
    "from helper import custom_put_along_axis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ea1921cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Minimal example\n",
    "reference = jnp.array([0])\n",
    "reads = jnp.array([[0], [1], [1], [1], [0], [1], [0], [1]])\n",
    "alphabet ='01'\n",
    "\n",
    "cluster_num = 5\n",
    "\n",
    "input_data = reference, reads, len(alphabet)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "744d2418",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0886777d",
   "metadata": {},
   "source": [
    "## Model - Finite DPM\n",
    "\n",
    "#### Finite Dirichlet mixture model with fixed number of clusters. \n",
    "\n",
    "ref = reference sequence  \n",
    "$N$ = number of reads  \n",
    "$K$ = number of clusters/components  \n",
    "$L$ = genome length (number of positions)  \n",
    "alphabet = {A, C, G, T, -}\n",
    "\n",
    "\n",
    "no-mutation rate: $\\gamma \\sim Beta(a,b)$   \n",
    "no-error rate: $\\theta \\sim Beta(c,d)$   \n",
    "Cluster weights ($K$-dim): $\\pi | \\alpha \\sim Dir(\\alpha)$  \n",
    "Cluster assignments ($N$-dim): $z|\\pi \\sim Categorical(\\pi)$  \n",
    "Cluster centers/haplotypes ($K$x$L$-dim): $h | ref, \\gamma \\sim Categorical(W) $ \n",
    "with $W(l,i)=\n",
    "\\begin{cases} \n",
    "\\gamma,  \\text{ if }i = ref[l] \\\\\n",
    "\\frac{1-\\gamma}{4}, \\text{ else. }\n",
    "\\end{cases}$ for $l \\in {1, ..., L}$ and $i\\in {1,..., |alphabet|}$  \n",
    "Likelihood of the reads ($N$-dim): $r | z, h, \\theta \\sim Categorical(E)$ \n",
    "with $E(n,l,i)=\n",
    "\\begin{cases} \n",
    "\\theta,  \\text{ if }i = h_{z_n}[l] \\\\\n",
    "\\frac{1-\\theta}{4}, \\text{ else. }\n",
    "\\end{cases}$ for $n \\in {1, ..., N}$, $l \\in {1, ..., L}$ and $i\\in {1,..., |alphabet|}$  \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "77ae49d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def model_finiteDPM(input_data):\n",
    "    reference, read_data, alphabet_length = input_data\n",
    "\n",
    "    # parameters\n",
    "    read_count = read_data.shape[0]\n",
    "    genome_length = read_data.shape[1]\n",
    "    alphabet_length = alphabet_length\n",
    "\n",
    "    # define rates\n",
    "    mutation_rate = numpyro.sample('mutation_rate', dist.Beta(1, 1))\n",
    "    error_rate = numpyro.sample('error_rate', dist.Beta(1, 1))\n",
    "\n",
    "    # create matrix of rates\n",
    "    mutation_rate_matrix = jnp.full(\n",
    "        (genome_length, alphabet_length), (1 - mutation_rate) / (alphabet_length - 1)\n",
    "    )\n",
    "    mutation_rate_matrix = custom_put_along_axis(\n",
    "        mutation_rate_matrix, reference.reshape(genome_length, 1), mutation_rate, axis=1\n",
    "    )\n",
    "\n",
    "    alpha0 = 0.001 # alpha must be bigger than zero\n",
    "    alpha = alpha0 * jnp.ones(cluster_num) / cluster_num\n",
    "    cluster_weights = numpyro.sample(\"cluster_weights\", dist.Dirichlet(jnp.ones(cluster_num)))\n",
    "\n",
    "    genome_axis = numpyro.plate(\"genome_axis\", genome_length, dim=-1)\n",
    "    with numpyro.plate(\"haplotype_axis\", cluster_num, dim=-2):\n",
    "        \n",
    "        with genome_axis:\n",
    "            haplotypes = numpyro.sample(\"haplotypes\", dist.Categorical(mutation_rate_matrix))  # cluster centers\n",
    "\n",
    "    with numpyro.plate(\"read_axis\", read_count, dim=-2):\n",
    "        cluster_assignments = numpyro.sample(\"cluster_assignments\", dist.Categorical(cluster_weights))\n",
    "        \n",
    "        with genome_axis:\n",
    "            \n",
    "            error_rate_matrix = jnp.full(\n",
    "                (read_count, genome_length, alphabet_length), (1 - error_rate) / (alphabet_length - 1)\n",
    "            )\n",
    "            error_rate_matrix = custom_put_along_axis(\n",
    "                error_rate_matrix, haplotypes[cluster_assignments].reshape(read_count, genome_length, 1), error_rate, axis=2\n",
    "            )\n",
    "\n",
    "            obs=numpyro.sample(\"obs\", dist.Categorical(error_rate_matrix), obs=read_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "72b832d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.50.0 (20211204.2007)\n",
       " -->\n",
       "<!-- Pages: 1 -->\n",
       "<svg width=\"442pt\" height=\"269pt\"\n",
       " viewBox=\"0.00 0.00 442.00 269.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 265)\">\n",
       "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-265 438,-265 438,4 -4,4\"/>\n",
       "<g id=\"clust1\" class=\"cluster\">\n",
       "<title>cluster_genome_axis</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"8,-8 8,-205 138,-205 138,-8 8,-8\"/>\n",
       "<text text-anchor=\"middle\" x=\"93.5\" y=\"-15.8\" font-family=\"Times,serif\" font-size=\"14.00\">genome_axis</text>\n",
       "</g>\n",
       "<g id=\"clust2\" class=\"cluster\">\n",
       "<title>cluster_haplotype_axis</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"16,-122 16,-197 130,-197 130,-122 16,-122\"/>\n",
       "<text text-anchor=\"middle\" x=\"81\" y=\"-129.8\" font-family=\"Times,serif\" font-size=\"14.00\">haplotype_axis</text>\n",
       "</g>\n",
       "<g id=\"clust3\" class=\"cluster\">\n",
       "<title>cluster_read_axis</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"60,-39 60,-114 130,-114 130,-39 60,-39\"/>\n",
       "<text text-anchor=\"middle\" x=\"96\" y=\"-46.8\" font-family=\"Times,serif\" font-size=\"14.00\">read_axis</text>\n",
       "</g>\n",
       "<g id=\"clust4\" class=\"cluster\">\n",
       "<title>cluster_read_axis__CLONE</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"246,-122 246,-197 426,-197 426,-122 246,-122\"/>\n",
       "<text text-anchor=\"middle\" x=\"392\" y=\"-129.8\" font-family=\"Times,serif\" font-size=\"14.00\">read_axis</text>\n",
       "</g>\n",
       "<!-- mutation_rate -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>mutation_rate</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"73\" cy=\"-243\" rx=\"59.59\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"73\" y=\"-239.3\" font-family=\"Times,serif\" font-size=\"14.00\">mutation_rate</text>\n",
       "</g>\n",
       "<!-- haplotypes -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>haplotypes</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"73\" cy=\"-171\" rx=\"48.99\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"73\" y=\"-167.3\" font-family=\"Times,serif\" font-size=\"14.00\">haplotypes</text>\n",
       "</g>\n",
       "<!-- mutation_rate&#45;&gt;haplotypes -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>mutation_rate&#45;&gt;haplotypes</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M73,-224.7C73,-216.98 73,-207.71 73,-199.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"76.5,-199.1 73,-189.1 69.5,-199.1 76.5,-199.1\"/>\n",
       "</g>\n",
       "<!-- error_rate -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>error_rate</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"191\" cy=\"-171\" rx=\"45.49\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"191\" y=\"-167.3\" font-family=\"Times,serif\" font-size=\"14.00\">error_rate</text>\n",
       "</g>\n",
       "<!-- obs -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>obs</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"95\" cy=\"-88\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"95\" y=\"-84.3\" font-family=\"Times,serif\" font-size=\"14.00\">obs</text>\n",
       "</g>\n",
       "<!-- error_rate&#45;&gt;obs -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>error_rate&#45;&gt;obs</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M172.49,-154.38C157.23,-141.5 135.47,-123.14 118.89,-109.16\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"121.12,-106.46 111.22,-102.68 116.6,-111.81 121.12,-106.46\"/>\n",
       "</g>\n",
       "<!-- cluster_weights -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>cluster_weights</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"336\" cy=\"-243\" rx=\"65.79\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"336\" y=\"-239.3\" font-family=\"Times,serif\" font-size=\"14.00\">cluster_weights</text>\n",
       "</g>\n",
       "<!-- cluster_assignments -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>cluster_assignments</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"336\" cy=\"-171\" rx=\"81.79\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"336\" y=\"-167.3\" font-family=\"Times,serif\" font-size=\"14.00\">cluster_assignments</text>\n",
       "</g>\n",
       "<!-- cluster_weights&#45;&gt;cluster_assignments -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>cluster_weights&#45;&gt;cluster_assignments</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M336,-224.7C336,-216.98 336,-207.71 336,-199.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"339.5,-199.1 336,-189.1 332.5,-199.1 339.5,-199.1\"/>\n",
       "</g>\n",
       "<!-- haplotypes&#45;&gt;obs -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>haplotypes&#45;&gt;obs</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M77.67,-152.82C80.6,-142.01 84.45,-127.84 87.78,-115.59\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"91.17,-116.44 90.42,-105.88 84.42,-114.61 91.17,-116.44\"/>\n",
       "</g>\n",
       "<!-- cluster_assignments&#45;&gt;obs -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>cluster_assignments&#45;&gt;obs</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M309.8,-153.75C292.07,-143.38 267.81,-130.35 245,-122 207.32,-108.2 162.11,-99.22 131.24,-94.17\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"131.77,-90.71 121.35,-92.6 130.68,-97.62 131.77,-90.71\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x7f8bc060e550>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "numpyro.render_model(model_finiteDPM, model_args=(input_data,))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06f54e72",
   "metadata": {},
   "source": [
    "-----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4aa5df1",
   "metadata": {},
   "source": [
    "## Model - Infinite DPM - Stickbreaking process\n",
    "#### Dirichlet mixture model where number of clusters is learn.\n",
    "\n",
    "ref = reference sequence  \n",
    "$N$ = number of reads  \n",
    "$K$ = number of clusters/components  \n",
    "$L$ = genome length (number of positions)  \n",
    "alphabet = {A, C, G, T, -}\n",
    "\n",
    "\n",
    "no-mutation rate: $\\gamma \\sim Beta(a,b)$   \n",
    "no-error rate: $\\theta \\sim Beta(c,d)$  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2f55b45d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mix_weights(beta):\n",
    "    beta1m_cumprod = (1 - beta).cumprod(-1)\n",
    "    return jnp.pad(beta, (0, 1), constant_values=1) * jnp.pad(\n",
    "        beta1m_cumprod, (1, 0), constant_values=1\n",
    "    )\n",
    "\n",
    "\n",
    "def model_infiniteSBP(input_data):\n",
    "    reference, read_data, alphabet_length = input_data\n",
    "\n",
    "    max_cluster_num_guess = 10  # guess for maximum number of clusters\n",
    "\n",
    "    # parameters\n",
    "    read_count = read_data.shape[0]\n",
    "    genome_length = reference.shape[0]\n",
    "    alphabet_length = alphabet_length #jnp.unique(reference).shape[0]\n",
    "\n",
    "    # define rates\n",
    "    mutation_rate = numpyro.sample(\"mutation_rate\", dist.Beta(1, 1))\n",
    "    error_rate = numpyro.sample(\"error_rate\", dist.Beta(1, 1))\n",
    "\n",
    "    # create matrix of rates\n",
    "    mutation_rate_matrix = jnp.full(\n",
    "        (genome_length, alphabet_length), (1 - mutation_rate) / (alphabet_length - 1)\n",
    "    )\n",
    "    mutation_rate_matrix = custom_put_along_axis(\n",
    "        mutation_rate_matrix, reference.reshape(genome_length, 1), mutation_rate, axis=1\n",
    "    )\n",
    "\n",
    "    with numpyro.plate(\"beta_plate\", max_cluster_num_guess - 1):\n",
    "        beta = numpyro.sample(\"beta\", dist.Beta(1, 1.1))\n",
    "    \n",
    "    cluster_weights = numpyro.sample(\n",
    "        \"cluster_weights\", dist.Dirichlet(mix_weights(beta))\n",
    "    )\n",
    "\n",
    "    genome_axis = numpyro.plate(\"genome_axis\", genome_length, dim=-1)\n",
    "    with numpyro.plate(\"haplotype_axis\", max_cluster_num_guess, dim=-2):\n",
    "        with genome_axis:\n",
    "            haplotypes = numpyro.sample(\n",
    "                \"haplotypes\", dist.Categorical(mutation_rate_matrix)\n",
    "            )  # cluster centers\n",
    "\n",
    "    with numpyro.plate(\"read_axis\", read_count, dim=-2):\n",
    "        cluster_assignments = numpyro.sample(\n",
    "            \"cluster_assignments\", dist.Categorical(cluster_weights)\n",
    "        )\n",
    "\n",
    "        with genome_axis:\n",
    "            error_rate_matrix = jnp.full(\n",
    "                (read_count, genome_length, alphabet_length),\n",
    "                (1 - error_rate) / (alphabet_length - 1),\n",
    "            )\n",
    "            error_rate_matrix = custom_put_along_axis(\n",
    "                error_rate_matrix,\n",
    "                haplotypes[cluster_assignments].reshape(read_count, genome_length, 1),\n",
    "                error_rate,\n",
    "                axis=2,\n",
    "            )\n",
    "\n",
    "            numpyro.sample(\"obs\", dist.Categorical(error_rate_matrix), obs=read_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3cf28e8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traced<ShapedArray(float32[9])>with<DynamicJaxprTrace(level=1/0)>\n",
      "Traced<ShapedArray(float32[10])>with<DynamicJaxprTrace(level=1/0)>\n",
      "model  (10,)\n",
      "Traced<ShapedArray(float32[9;_provenance:frozenset({'beta'})])>with<DynamicJaxprTrace(level=1/0)>\n",
      "Traced<ShapedArray(float32[10;_provenance:frozenset({'beta'})])>with<_ProvenanceJaxprTrace(level=1/0)>\n",
      "model  (10,)\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.50.0 (20211204.2007)\n",
       " -->\n",
       "<!-- Pages: 1 -->\n",
       "<svg width=\"442pt\" height=\"360pt\"\n",
       " viewBox=\"0.00 0.00 442.00 360.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 356)\">\n",
       "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-356 438,-356 438,4 -4,4\"/>\n",
       "<g id=\"clust1\" class=\"cluster\">\n",
       "<title>cluster_beta_plate</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"300,-269 300,-344 372,-344 372,-269 300,-269\"/>\n",
       "<text text-anchor=\"middle\" x=\"336\" y=\"-276.8\" font-family=\"Times,serif\" font-size=\"14.00\">beta_plate</text>\n",
       "</g>\n",
       "<g id=\"clust2\" class=\"cluster\">\n",
       "<title>cluster_genome_axis</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"8,-8 8,-205 138,-205 138,-8 8,-8\"/>\n",
       "<text text-anchor=\"middle\" x=\"93.5\" y=\"-15.8\" font-family=\"Times,serif\" font-size=\"14.00\">genome_axis</text>\n",
       "</g>\n",
       "<g id=\"clust3\" class=\"cluster\">\n",
       "<title>cluster_haplotype_axis</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"16,-122 16,-197 130,-197 130,-122 16,-122\"/>\n",
       "<text text-anchor=\"middle\" x=\"81\" y=\"-129.8\" font-family=\"Times,serif\" font-size=\"14.00\">haplotype_axis</text>\n",
       "</g>\n",
       "<g id=\"clust4\" class=\"cluster\">\n",
       "<title>cluster_read_axis</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"60,-39 60,-114 130,-114 130,-39 60,-39\"/>\n",
       "<text text-anchor=\"middle\" x=\"96\" y=\"-46.8\" font-family=\"Times,serif\" font-size=\"14.00\">read_axis</text>\n",
       "</g>\n",
       "<g id=\"clust5\" class=\"cluster\">\n",
       "<title>cluster_read_axis__CLONE</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"246,-122 246,-197 426,-197 426,-122 246,-122\"/>\n",
       "<text text-anchor=\"middle\" x=\"392\" y=\"-129.8\" font-family=\"Times,serif\" font-size=\"14.00\">read_axis</text>\n",
       "</g>\n",
       "<!-- mutation_rate -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>mutation_rate</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"73\" cy=\"-243\" rx=\"59.59\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"73\" y=\"-239.3\" font-family=\"Times,serif\" font-size=\"14.00\">mutation_rate</text>\n",
       "</g>\n",
       "<!-- haplotypes -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>haplotypes</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"73\" cy=\"-171\" rx=\"48.99\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"73\" y=\"-167.3\" font-family=\"Times,serif\" font-size=\"14.00\">haplotypes</text>\n",
       "</g>\n",
       "<!-- mutation_rate&#45;&gt;haplotypes -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>mutation_rate&#45;&gt;haplotypes</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M73,-224.7C73,-216.98 73,-207.71 73,-199.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"76.5,-199.1 73,-189.1 69.5,-199.1 76.5,-199.1\"/>\n",
       "</g>\n",
       "<!-- error_rate -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>error_rate</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"191\" cy=\"-171\" rx=\"45.49\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"191\" y=\"-167.3\" font-family=\"Times,serif\" font-size=\"14.00\">error_rate</text>\n",
       "</g>\n",
       "<!-- obs -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>obs</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"95\" cy=\"-88\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"95\" y=\"-84.3\" font-family=\"Times,serif\" font-size=\"14.00\">obs</text>\n",
       "</g>\n",
       "<!-- error_rate&#45;&gt;obs -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>error_rate&#45;&gt;obs</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M172.49,-154.38C157.23,-141.5 135.47,-123.14 118.89,-109.16\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"121.12,-106.46 111.22,-102.68 116.6,-111.81 121.12,-106.46\"/>\n",
       "</g>\n",
       "<!-- cluster_weights -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>cluster_weights</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"336\" cy=\"-243\" rx=\"65.79\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"336\" y=\"-239.3\" font-family=\"Times,serif\" font-size=\"14.00\">cluster_weights</text>\n",
       "</g>\n",
       "<!-- cluster_assignments -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>cluster_assignments</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"336\" cy=\"-171\" rx=\"81.79\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"336\" y=\"-167.3\" font-family=\"Times,serif\" font-size=\"14.00\">cluster_assignments</text>\n",
       "</g>\n",
       "<!-- cluster_weights&#45;&gt;cluster_assignments -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>cluster_weights&#45;&gt;cluster_assignments</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M336,-224.7C336,-216.98 336,-207.71 336,-199.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"339.5,-199.1 336,-189.1 332.5,-199.1 339.5,-199.1\"/>\n",
       "</g>\n",
       "<!-- beta -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>beta</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"336\" cy=\"-318\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"336\" y=\"-314.3\" font-family=\"Times,serif\" font-size=\"14.00\">beta</text>\n",
       "</g>\n",
       "<!-- beta&#45;&gt;cluster_weights -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>beta&#45;&gt;cluster_weights</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M336,-299.7C336,-291.25 336,-280.87 336,-271.37\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"339.5,-271.18 336,-261.18 332.5,-271.18 339.5,-271.18\"/>\n",
       "</g>\n",
       "<!-- haplotypes&#45;&gt;obs -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>haplotypes&#45;&gt;obs</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M77.67,-152.82C80.6,-142.01 84.45,-127.84 87.78,-115.59\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"91.17,-116.44 90.42,-105.88 84.42,-114.61 91.17,-116.44\"/>\n",
       "</g>\n",
       "<!-- cluster_assignments&#45;&gt;obs -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>cluster_assignments&#45;&gt;obs</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M309.8,-153.75C292.07,-143.38 267.81,-130.35 245,-122 207.32,-108.2 162.11,-99.22 131.24,-94.17\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"131.77,-90.71 121.35,-92.6 130.68,-97.62 131.77,-90.71\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x7f8bc062c250>"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "numpyro.render_model(model_infiniteSBP, model_args=(input_data,))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c78d787c",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56ca44c4",
   "metadata": {},
   "source": [
    "## Model - Finite DPM\n",
    "\n",
    "#### Finite Dirichlet mixture model with fixed number of clusters. With extension of error and mutation model accounting for number of matches, number of subsitutions, number of deletions\n",
    "\n",
    "ref = reference sequence  \n",
    "$N$ = number of reads  \n",
    "$K$ = number of clusters/components  \n",
    "$L$ = genome length (number of positions)  \n",
    "alphabet = {A, C, G, T, -}\n",
    "\n",
    "\n",
    "mutation rate: $\\gamma, \\gamma_S, \\gamma_{D^f} \\gamma_{D^b}\\sim Dirichlet(a,b,c,d)$  \n",
    "\n",
    "mutation rate matrix depending on the reference:\n",
    "$\n",
    "\\begin{matrix}\n",
    "ref & A & C & G & T & - \\\\\n",
    "A & \\gamma & \\gamma_S & \\gamma_S & \\gamma_S & \\gamma_{D^b} \\\\\n",
    "C & \\gamma_S & \\gamma & \\gamma_S & \\gamma_S & \\gamma_{D^b} \\\\\n",
    "G & \\gamma_S & \\gamma_S & \\gamma & \\gamma_S & \\gamma_{D^b} \\\\\n",
    "T & \\gamma_S & \\gamma_S & \\gamma_S & \\gamma & \\gamma_{D^b} \\\\\n",
    "- & \\gamma_{D^f} & \\gamma_{D^f}& \\gamma_{D^f} & \\gamma_{D^f} & \\gamma \\\\\n",
    "\\end{matrix}\n",
    "$\n",
    "\n",
    "error rate: $\\theta, \\theta_S, \\theta_{D^f} \\theta_{D^b} \\sim Dirichlet(a,b,c,d)$   \n",
    "\n",
    "error rate matrix depending on the assigned haplotype:\n",
    "$\n",
    "\\begin{matrix}\n",
    "haplotype & A & C & G & T & - \\\\\n",
    "A & \\theta & \\theta_S & \\theta_S & \\theta_S & \\theta_{D^b} \\\\\n",
    "C & \\theta_S & \\theta & \\theta_S & \\theta_S & \\theta_{D^b} \\\\\n",
    "G & \\theta_S & \\theta_S & \\theta & \\theta_S & \\theta_{D^b} \\\\\n",
    "T & \\theta_S & \\theta_S & \\theta_S & \\theta & \\theta_{D^b} \\\\\n",
    "- & \\theta_{D^f} & \\theta_{D^f}& \\theta_{D^f} & \\theta_{D^f} & \\theta \\\\\n",
    "\\end{matrix}\n",
    "$\n",
    "\n",
    "\n",
    "Cluster weights ($K$-dim): $\\pi | \\alpha \\sim Dir(\\alpha)$ \n",
    "\n",
    "Cluster assignments ($N$-dim): $z|\\pi \\sim Categorical(\\pi)$  \n",
    "\n",
    "Cluster centers/haplotypes ($K$x$L$-dim): $h | ref, \\gamma \\sim Categorical(W) $ \n",
    "with $W(l,i)=$ mutation rates (see above).\n",
    "\n",
    "Likelihood of the reads ($N$-dim): $r | z, h, \\theta \\sim Categorical(E)$ \n",
    "with $E(n,l,i)=$  error rate matrix (see above).\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "19d99336",
   "metadata": {},
   "outputs": [],
   "source": [
    "def model_finiteDPM_extended(input_data):\n",
    "    '''\n",
    "    This model is only for alphabet={A,C,G,T,-} as deletion mutations and errors are treated seperatly \n",
    "    (wrt to subsitutions).\n",
    "    '''\n",
    "    cluster_num = 10\n",
    "    reference, read_data, alphabet_length = input_data\n",
    "\n",
    "    # parameters\n",
    "    read_count = read_data.shape[0]\n",
    "    genome_length = read_data.shape[1]\n",
    "    alphabet_length = alphabet_length\n",
    "\n",
    "    # define rates\n",
    "    mutation_rate = numpyro.sample('mutation_rate', dist.Dirichlet(jnp.ones(4)/4))\n",
    "    error_rate = numpyro.sample('error_rate', dist.Dirichlet(jnp.ones(4)/4))\n",
    "\n",
    "    # create matrix of rates\n",
    "    mutation_rate_matrix = jnp.full(\n",
    "        (genome_length, alphabet_length), mutation_rate[1] / (alphabet_length - 2)\n",
    "    ) # subsitutions\n",
    "    mutation_rate_matrix = mutation_rate_matrix.at[:,4].set(mutation_rate[2]) # deletion in haplotype\n",
    "    \n",
    "    '''\n",
    "    # TODO: \n",
    "    # Option 0: \n",
    "    mutation_rate_matrix = mutation_rate_matrix.at[indicies,:].set(mutation_rate[3] / (alphabet_length - 1))\n",
    "    # Option 1: \n",
    "    mutation_rate_matrix = mutation_rate_matrix.at[indicies,:].set((mutation_rate[3] / (alphabet_length - 1))*jnp.ones(5))\n",
    "    # Option 2: \n",
    "    indicies = jnp.argwhere(reference==4)\n",
    "    mutation_rate_matrix = mutation_rate_matrix.at[indicies,0].set(\n",
    "        mutation_rate[3] / (alphabet_length - 1)\n",
    "    )\n",
    "    mutation_rate_matrix = mutation_rate_matrix.at[indicies,1].set(\n",
    "        mutation_rate[3] / (alphabet_length - 1)\n",
    "    )\n",
    "    mutation_rate_matrix = mutation_rate_matrix.at[indicies,2].set(\n",
    "        mutation_rate[3] / (alphabet_length - 1)\n",
    "    )\n",
    "    mutation_rate_matrix = mutation_rate_matrix.at[indicies,3].set(\n",
    "        mutation_rate[3] / (alphabet_length - 1)\n",
    "    )\n",
    "    mutation_rate_matrix = mutation_rate_matrix.at[indicies,4].set(\n",
    "        mutation_rate[0]\n",
    "    )\n",
    "    '''\n",
    "    mutation_rate_matrix = custom_put_along_axis(\n",
    "        mutation_rate_matrix, reference.reshape(genome_length, 1), mutation_rate[0], axis=1\n",
    "    ) # matches\n",
    "    mutation_rate_matrix = mutation_rate_matrix / mutation_rate_matrix.sum(axis=1)[:,jnp.newaxis] #normalize\n",
    "    \n",
    "\n",
    "    alpha0 = 0.001 # alpha must be bigger than zero\n",
    "    alpha = alpha0 * jnp.ones(cluster_num) / cluster_num\n",
    "    cluster_weights = numpyro.sample(\"cluster_weights\", dist.Dirichlet(jnp.ones(cluster_num)))\n",
    "\n",
    "    genome_axis = numpyro.plate(\"genome_axis\", genome_length, dim=-1)\n",
    "    with numpyro.plate(\"haplotype_axis\", cluster_num, dim=-2):\n",
    "        \n",
    "        with genome_axis:\n",
    "            haplotypes = numpyro.sample(\"haplotypes\", dist.Categorical(mutation_rate_matrix))  # cluster centers\n",
    "    \n",
    "    with numpyro.plate(\"read_axis\", read_count, dim=-2):\n",
    "        cluster_assignments = numpyro.sample(\"cluster_assignments\", dist.Categorical(cluster_weights))\n",
    "        with genome_axis:\n",
    "            error_rate_matrix = jnp.full(\n",
    "                (read_count, genome_length, alphabet_length),\n",
    "                error_rate[1] / (alphabet_length - 1),\n",
    "            ) # substitutions \n",
    "            error_rate_matrix = error_rate_matrix.at[:,:,4].set(error_rate[2]) # deletion in read\n",
    "            '''\n",
    "            # TODO: Find solution to this problem!! It is doing the right thing but its messing up \n",
    "            # after one or two iterations\n",
    "            # deletion in haplotype:\n",
    "            # this are the indices where we want to have theta_Db\n",
    "            ind_list = jnp.argwhere(haplotypes[cluster_assignments].reshape(read_count, genome_length)==4)\n",
    "            for ind in ind_list:\n",
    "                error_rate_matrix = error_rate_matrix.at[ind[0],ind[1]].set(error_rate[3]) # deletion in haplotype \n",
    "            '''\n",
    "            error_rate_matrix = custom_put_along_axis(\n",
    "                error_rate_matrix,\n",
    "                haplotypes[cluster_assignments].reshape(read_count, genome_length, 1),\n",
    "                error_rate[0],\n",
    "                axis=2,\n",
    "            ) # matches \n",
    "\n",
    "            numpyro.sample(\"obs\", dist.Categorical(error_rate_matrix), obs=read_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "be2cc9b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.50.0 (20211204.2007)\n",
       " -->\n",
       "<!-- Pages: 1 -->\n",
       "<svg width=\"442pt\" height=\"269pt\"\n",
       " viewBox=\"0.00 0.00 442.00 269.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 265)\">\n",
       "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-265 438,-265 438,4 -4,4\"/>\n",
       "<g id=\"clust1\" class=\"cluster\">\n",
       "<title>cluster_genome_axis</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"8,-8 8,-205 138,-205 138,-8 8,-8\"/>\n",
       "<text text-anchor=\"middle\" x=\"93.5\" y=\"-15.8\" font-family=\"Times,serif\" font-size=\"14.00\">genome_axis</text>\n",
       "</g>\n",
       "<g id=\"clust2\" class=\"cluster\">\n",
       "<title>cluster_haplotype_axis</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"16,-122 16,-197 130,-197 130,-122 16,-122\"/>\n",
       "<text text-anchor=\"middle\" x=\"81\" y=\"-129.8\" font-family=\"Times,serif\" font-size=\"14.00\">haplotype_axis</text>\n",
       "</g>\n",
       "<g id=\"clust3\" class=\"cluster\">\n",
       "<title>cluster_read_axis</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"60,-39 60,-114 130,-114 130,-39 60,-39\"/>\n",
       "<text text-anchor=\"middle\" x=\"96\" y=\"-46.8\" font-family=\"Times,serif\" font-size=\"14.00\">read_axis</text>\n",
       "</g>\n",
       "<g id=\"clust4\" class=\"cluster\">\n",
       "<title>cluster_read_axis__CLONE</title>\n",
       "<polygon fill=\"none\" stroke=\"black\" points=\"246,-122 246,-197 426,-197 426,-122 246,-122\"/>\n",
       "<text text-anchor=\"middle\" x=\"392\" y=\"-129.8\" font-family=\"Times,serif\" font-size=\"14.00\">read_axis</text>\n",
       "</g>\n",
       "<!-- mutation_rate -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>mutation_rate</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"73\" cy=\"-243\" rx=\"59.59\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"73\" y=\"-239.3\" font-family=\"Times,serif\" font-size=\"14.00\">mutation_rate</text>\n",
       "</g>\n",
       "<!-- haplotypes -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>haplotypes</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"73\" cy=\"-171\" rx=\"48.99\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"73\" y=\"-167.3\" font-family=\"Times,serif\" font-size=\"14.00\">haplotypes</text>\n",
       "</g>\n",
       "<!-- mutation_rate&#45;&gt;haplotypes -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>mutation_rate&#45;&gt;haplotypes</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M73,-224.7C73,-216.98 73,-207.71 73,-199.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"76.5,-199.1 73,-189.1 69.5,-199.1 76.5,-199.1\"/>\n",
       "</g>\n",
       "<!-- error_rate -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>error_rate</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"191\" cy=\"-171\" rx=\"45.49\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"191\" y=\"-167.3\" font-family=\"Times,serif\" font-size=\"14.00\">error_rate</text>\n",
       "</g>\n",
       "<!-- obs -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>obs</title>\n",
       "<ellipse fill=\"grey\" stroke=\"black\" cx=\"95\" cy=\"-88\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"95\" y=\"-84.3\" font-family=\"Times,serif\" font-size=\"14.00\">obs</text>\n",
       "</g>\n",
       "<!-- error_rate&#45;&gt;obs -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>error_rate&#45;&gt;obs</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M172.49,-154.38C157.23,-141.5 135.47,-123.14 118.89,-109.16\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"121.12,-106.46 111.22,-102.68 116.6,-111.81 121.12,-106.46\"/>\n",
       "</g>\n",
       "<!-- cluster_weights -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>cluster_weights</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"336\" cy=\"-243\" rx=\"65.79\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"336\" y=\"-239.3\" font-family=\"Times,serif\" font-size=\"14.00\">cluster_weights</text>\n",
       "</g>\n",
       "<!-- cluster_assignments -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>cluster_assignments</title>\n",
       "<ellipse fill=\"white\" stroke=\"black\" cx=\"336\" cy=\"-171\" rx=\"81.79\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"336\" y=\"-167.3\" font-family=\"Times,serif\" font-size=\"14.00\">cluster_assignments</text>\n",
       "</g>\n",
       "<!-- cluster_weights&#45;&gt;cluster_assignments -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>cluster_weights&#45;&gt;cluster_assignments</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M336,-224.7C336,-216.98 336,-207.71 336,-199.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"339.5,-199.1 336,-189.1 332.5,-199.1 339.5,-199.1\"/>\n",
       "</g>\n",
       "<!-- haplotypes&#45;&gt;obs -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>haplotypes&#45;&gt;obs</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M77.67,-152.82C80.6,-142.01 84.45,-127.84 87.78,-115.59\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"91.17,-116.44 90.42,-105.88 84.42,-114.61 91.17,-116.44\"/>\n",
       "</g>\n",
       "<!-- cluster_assignments&#45;&gt;obs -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>cluster_assignments&#45;&gt;obs</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M309.8,-153.75C292.07,-143.38 267.81,-130.35 245,-122 207.32,-108.2 162.11,-99.22 131.24,-94.17\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"131.77,-90.71 121.35,-92.6 130.68,-97.62 131.77,-90.71\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x7fc48190ad30>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "numpyro.render_model(model_finiteDPM_extended, model_args=(input_data,))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "430920af",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
