{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6eeec1d7",
   "metadata": {},
   "source": [
    "## Model - Infinite DPM - Chinese Restaurant Mixture Model (CRPMM)\n",
    "\n",
    "#### Dirichlet mixture model where number of clusters is learned. \n",
    "\n",
    "ref = reference sequence  \n",
    "$N$ = number of reads  \n",
    "$K$ = number of clusters/components  \n",
    "$L$ = genome length (number of positions)  \n",
    "alphabet = {A, C, G, T, -}\n",
    "\n",
    "\n",
    "no-mutation rate: $\\gamma \\sim Beta(a,b)$   \n",
    "no-error rate: $\\theta \\sim Beta(c,d)$   \n",
    "Cluster weights ($K$-dim): $\\pi | \\alpha \\sim Dir(\\alpha)$  \n",
    "Cluster assignments ($N$-dim): $z|\\pi \\sim Categorical(\\pi)$  \n",
    "Cluster centers/haplotypes ($K$x$L$-dim): $h | ref, \\gamma \\sim Categorical(W) $ \n",
    "with $W(l,i)=\n",
    "\\begin{cases} \n",
    "\\gamma,  \\text{ if }i = ref[l] \\\\\n",
    "\\frac{1-\\gamma}{4}, \\text{ else. }\n",
    "\\end{cases}$ for $l \\in {1, ..., L}$ and $i\\in {1,..., |alphabet|}$  \n",
    "Likelihood of the reads ($N$-dim): $r | z, h, \\theta \\sim Categorical(E)$ \n",
    "with $E(n,l,i)=\n",
    "\\begin{cases} \n",
    "\\theta,  \\text{ if }i = h_{z_n}[l] \\\\\n",
    "\\frac{1-\\theta}{4}, \\text{ else. }\n",
    "\\end{cases}$ for $n \\in {1, ..., N}$, $l \\in {1, ..., L}$ and $i\\in {1,..., |alphabet|}$  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "00392990",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpyro\n",
    "import numpyro.distributions as dist\n",
    "from numpyro.infer import MCMC, NUTS, DiscreteHMCGibbs, Predictive\n",
    "\n",
    "from jax import random\n",
    "import jax\n",
    "import jax.numpy as jnp\n",
    "\n",
    "import arviz as az\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a9b3c9f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Minimal example\n",
    "reference = jnp.array([0])\n",
    "reads = jnp.array([[0], [1], [1], [1], [0], [1], [0], [1]])\n",
    "alphabet ='01'\n",
    "\n",
    "cluster_num = 5\n",
    "\n",
    "input_data = reference, reads, len(alphabet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ed27747d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the following as inspiration\n",
    "# https://forum.pyro.ai/t/variational-inference-for-dirichlet-process-clustering/98/2 \n",
    "\n",
    "def model_infiniteCRPMM(input_data):\n",
    "    reference, read_data, alphabet_length = input_data\n",
    "\n",
    "    # parameters\n",
    "    read_count = read_data.shape[0]\n",
    "    genome_length = read_data.shape[1]\n",
    "    alphabet_length = alphabet_length\n",
    "    alpha0 = 0.1 \n",
    "    \n",
    "    haplotypes = {}  # sample this lazily\n",
    "    crp_counts = [] \n",
    "\n",
    "    # define rates\n",
    "    mutation_rate = numpyro.sample('mutation_rate', dist.Beta(1, 1))\n",
    "    error_rate = numpyro.sample('error_rate', dist.Beta(1, 1))\n",
    "\n",
    "    # create matrix of rates\n",
    "    mutation_rate_matrix = jnp.full((genome_length, alphabet_length), (1 - mutation_rate) / (alphabet_length - 1))\n",
    "    mutation_rate_matrix = custom_put_along_axis(mutation_rate_matrix, reference.reshape(genome_length, 1), mutation_rate, axis=1)\n",
    "\n",
    "    #loc, scale = jnp.zeros(1), jnp.ones(1)*2\n",
    "    #alpha = numpyro.sample(\"alpha\", dist.LogNormal(loc,scale)) # alpha must be more than zero\n",
    "    \n",
    "    for n in range(read_count):\n",
    "        print('----')\n",
    "        print('read number ', n)\n",
    "        print('crp_counts ', crp_counts)\n",
    "        # sample from a CRP\n",
    "        weights = jnp.array(crp_counts + [alpha0])\n",
    "        weights /= weights.sum()\n",
    "        print('weights ', weights)\n",
    "        \n",
    "        cluster_assignments = numpyro.sample(\"cluster_assignments\"+str(n), dist.Categorical(weights))\n",
    "        print('cluster_assignments', cluster_assignments)\n",
    "        \n",
    "        if cluster_assignments >= len(crp_counts): \n",
    "            # new cluster \n",
    "            crp_counts.append(1) \n",
    "        else: \n",
    "            # cluster already exists\n",
    "            crp_counts[cluster_assignments] += 1\n",
    "            \n",
    "        # sample haplotypes\n",
    "        # lazily sample cluster mean\n",
    "        if int(cluster_assignments) not in haplotypes.keys():\n",
    "            haplotypes[int(cluster_assignments)] = numpyro.sample(\"haplotypes\"+str(cluster_assignments), dist.Categorical(mutation_rate_matrix))\n",
    "        \n",
    "        print('shape haplotypes[int(cluster_assignments)] ', haplotypes[int(cluster_assignments)].shape)\n",
    "        error_rate_matrix = jnp.full((genome_length, alphabet_length), (1 - error_rate) / (alphabet_length - 1))\n",
    "        print('error_rate ', error_rate)\n",
    "        print('shape error_rate_matrix', error_rate_matrix.shape)\n",
    "        print('before ' , type(error_rate_matrix))\n",
    "        print('haplotypes[int(cluster_assignments)] ',haplotypes[int(cluster_assignments)])\n",
    "        error_rate_matrix = custom_put_along_axis(error_rate_matrix, haplotypes[int(cluster_assignments)].reshape(genome_length, 1), error_rate, axis=1)\n",
    "        \n",
    "        print('after ',type(error_rate_matrix))\n",
    "        \n",
    "        obs = numpyro.sample(\"obs\"+str(n), dist.Categorical(error_rate_matrix), obs=read_data[n])\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e899447",
   "metadata": {},
   "outputs": [],
   "source": [
    "rng_key = jax.random.PRNGKey(0)\n",
    "\n",
    "num_warmup, num_samples = 2000, 20000\n",
    "\n",
    "model = model_infiniteCRPMM\n",
    "\n",
    "# Run NUTS. How many chains? \n",
    "kernel = NUTS(model)\n",
    "mcmc = MCMC(\n",
    "    DiscreteHMCGibbs(kernel),\n",
    "    num_warmup=num_warmup,\n",
    "    num_samples=num_samples,\n",
    "    num_chains=2\n",
    ")\n",
    "mcmc.run(rng_key, input_data)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
